.cbmc-batch/jobs/aws_byte_cursor_compare_lexical/aws_byte_cursor_compare_lexical_harness.c

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/byte_buf.h>
+#include <proof_helpers/make_common_data_structures.h>
+#include <proof_helpers/proof_allocators.h>
+
+void aws_byte_cursor_compare_lexical_harness() {
+    /* parameters */
+    struct aws_byte_cursor lhs;
+    struct aws_byte_cursor rhs;
+
+    /* assumptions */
+    __CPROVER_assume(aws_byte_cursor_is_bounded(&lhs, MAX_BUFFER_SIZE));
+    ensure_byte_cursor_has_allocated_buffer_member(&lhs);
+    __CPROVER_assume(aws_byte_cursor_is_valid(&lhs));
+    if (nondet_bool()) {
+        rhs = lhs;
+    } else {
+        __CPROVER_assume(aws_byte_cursor_is_bounded(&rhs, MAX_BUFFER_SIZE));
+        ensure_byte_cursor_has_allocated_buffer_member(&rhs);
+        __CPROVER_assume(aws_byte_cursor_is_valid(&rhs));
+    }
+
+    /* save current state of the data structure */
+    struct aws_byte_cursor old_lhs = lhs;
+    struct store_byte_from_buffer old_byte_from_lhs;
+    save_byte_from_array(lhs.ptr, lhs.len, &old_byte_from_lhs);
+    struct aws_byte_cursor old_rhs = rhs;
+    struct store_byte_from_buffer old_byte_from_rhs;
+    save_byte_from_array(rhs.ptr, rhs.len, &old_byte_from_rhs);
+
+    /* operation under verification */
+    if (aws_byte_cursor_compare_lexical((nondet_bool() ? &lhs : NULL), (nondet_bool() ? &rhs : NULL)) == 0) {
+        assert(lhs.len == rhs.len);
+        if (lhs.len > 0) {
+            assert_bytes_match(lhs.ptr, rhs.ptr, lhs.len);
+        }
+    }
+
+    /* assertions */
+    assert(aws_byte_cursor_compare_lexical(&lhs, &lhs) == 0);
```

Comment:
This seems like an "operation under verification" rather than an "assertion". I suggest that you move it further up, so that it's associated with the other call to `aws_byte_cursor_compare_lexical`.

Also, I don't think it's okay to do
```c
aws_byte_cursor_compare_lexical(&lhs, &rhs);
aws_byte_cursor_compare_lexical(&lhs, &lhs);
assert(aws_byte_cursor_is_valid(&lhs));
```
because what if the first call to `aws_byte_cursor_compare_lexical` makes `lhs` invalid, but the second one makes it valid again? The harness doesn't catch this case. I think we should check for validity immediately after we do *each* of the function calls.

My suggestion is that you only make a single call to `aws_byte_cursor_compare_lexical` which deals with both identical and distinct `lhs` and `rhs`. Suggest:
```c
if (aws_byte_cursor_compare_lexical(
      (nondet_bool() ? &lhs : NULL),
      (nondet_bool() ? &lhs
                     : nondet_bool() ? rhs
                                     : NULL)) == 0) {
```

Reply:
That's great! However, keep in mind that in many cases now we should not pass `NULL` pointers to entry points in proofs. Since all `AWS_PRECONDITIONS` turned into assertions, we need to keep the assumptions in the proof harness; therefore, one of the assumptions here is that `lhs` and `rhs` must be different than NULL.

Reply:
Actually, I want to keep both calls otherwise I won't be able to check the following assertions when calling `aws_byte_cursor_compare_lexical(&lhs, &lhs);`

```
        assert(lhs.len == rhs.len);
        if (lhs.len > 0) {
            assert_bytes_match(lhs.ptr, rhs.ptr, lhs.len);
        }
```

In addition, the first call cannot make `lhs` or `rhs` invalid, because we ensure their validity in the postconditions.